core/src/main/java/org/mskcc/cbio/portal/scripts/ImportResourceData.java

```diff
@@ -0,0 +1,618 @@
+        int internalSampleId = -1;
+        int internalPatientId = -1;
+
+        // create resource_id set
+        Set<String> patientResourceIdSet = resources.stream().filter(resource -> resource.getResourceType().equals(ResourceDefinition.PATIENT_RESOURCE_TYPE)).map(resource -> resource.getResourceId()).collect(Collectors.toSet());
+        Set<String> sampleResourceIdSet = resources.stream().filter(resource -> resource.getResourceType().equals(ResourceDefinition.SAMPLE_RESOURCE_TYPE)).map(resource -> resource.getResourceId()).collect(Collectors.toSet());
+        Set<String> studyResourceIdSet = resources.stream().filter(resource -> resource.getResourceType().equals(ResourceDefinition.STUDY_RESOURCE_TYPE)).map(resource -> resource.getResourceId()).collect(Collectors.toSet());
+
+        // check if sample is not already added:
+        Sample sample = DaoSample.getSampleByCancerStudyAndSampleId(cancerStudy.getInternalId(), stableSampleId, false);
+        if (sample != null) {
+            // get internal sample id if sample exists
+            internalSampleId = sample.getInternalId();
+        } else {
+            Patient patient = DaoPatient.getPatientByCancerStudyAndPatientId(cancerStudy.getInternalId(),
+                    stablePatientId);
+            if (patient != null) {
+                // patient exists, get internal id:
+                internalPatientId = patient.getInternalId();
+            } else {
+                // add patient:
+                internalPatientId = (patientIdIndex >= 0) ? addPatientToDatabase(fields[patientIdIndex]) : -1;
+            }
+            // sample is new, so attempt to add to DB
+            internalSampleId = (stableSampleId.length() > 0) ? addSampleToDatabase(stableSampleId, fields, headerNames)
+                    : -1;
+        }
+
+        // validate and count:
+        if (internalSampleId != -1) {
+            // some minimal validation/fail safe for now: only continue if patientId is same
+            // as patient id in
+            // existing sample (can occur in case of this.isSupplementalData or in case of
+            // parsing bug in addSampleToDatabase):
+            internalPatientId = DaoPatient
+                    .getPatientByCancerStudyAndPatientId(cancerStudy.getInternalId(), stablePatientId).getInternalId();
+            if (internalPatientId != DaoSample.getSampleById(internalSampleId).getInternalPatientId()) {
+                throw new RuntimeException("Error: Sample " + stableSampleId
+                        + " was previously linked to another patient, and not to " + stablePatientId);
+            }
+            numSamplesProcessed++;
+        }
+
+        // if the resource id or url matches one of the missing values, skip this resource:
+        if ((resourceIdIndex != -1 && MissingResourceValues.has(fields[resourceIdIndex])) || (urlIndex != -1 && MissingResourceValues.has(fields[urlIndex]))) {
+            numEmptyResourcesSkipped++;
+        } else {
```

Comment:
this `else` block seems to complex to review

Reply:
@onursumer I updated the solution, could you take a look if that's better?

Reply:
@dippindots looks better, thanks!